ffi/src/repr_c.rs

```diff
@@ -0,0 +1,191 @@
+        while let Some((first, rest)) = substore.split_first_mut() {
+            store
+                .0
+                .push(FfiType::into_ffi(Clone::clone(&self[i]), first));
+            substore = rest;
+            i += 1;
+        }
+
+        SliceRef::from_slice(&store.0)
+    }
+
+    unsafe fn try_from_repr_c(
+        source: Self::ReprC,
+        store: &'itm mut Self::FfiStore,
+    ) -> Result<Self> {
+        let source = source.into_rust().ok_or(FfiReturn::ArgIsNull)?;
+        store.1 = core::iter::repeat_with(Default::default).collect();
+
+        let (mut i, mut substore) = (0, &mut store.1[..]);
+        while let Some((first, rest)) = substore.split_first_mut() {
+            store.0.push(FfiType::try_from_ffi(source[i], first)?);
+            substore = rest;
+            i += 1;
+        }
+
+        Ok(&store.0)
+    }
+}
+
+impl<'slice, T: Transmute> CType<'slice> for &'slice [Transmutable<T, T::ReprC>] {
+    type ReprC = SliceRef<T::ReprC>;
+    type RustStore = ();
+    type FfiStore = ();
+
+    fn into_repr_c(self, _: &mut ()) -> Self::ReprC {
+        let (ptr, len) = (self.as_ptr(), self.len());
+        SliceRef::from_raw_parts(ptr.cast(), len)
+    }
+    unsafe fn try_from_repr_c(source: Self::ReprC, _: &mut ()) -> Result<Self> {
+        let (ptr, len) = (source.as_ptr(), source.len());
+
+        if !source
+            .into_rust()
+            .ok_or(FfiReturn::ArgIsNull)?
+            .iter()
+            .all(|item| FfiReturn::Ok == T::is_valid(*item))
+        {
+            return Err(FfiReturn::TrapRepresentation);
+        }
+
+        Ok(core::slice::from_raw_parts(ptr.cast(), len))
+    }
+}
+
+impl<'slice, T: Transmute> CType<'slice> for &'slice mut [Transmutable<T, T::ReprC>] {
+    type ReprC = SliceMut<T::ReprC>;
+    type RustStore = ();
+    type FfiStore = ();
+
+    fn into_repr_c(self, _: &mut ()) -> Self::ReprC {
+        let (ptr, len) = (self.as_mut_ptr(), self.len());
+        SliceMut::from_raw_parts(ptr.cast(), len)
+    }
+    unsafe fn try_from_repr_c(source: Self::ReprC, _: &mut ()) -> Result<Self> {
+        let (ptr, len) = (source.as_mut_ptr(), source.len());
+
+        if !source
+            .into_rust()
+            .ok_or(FfiReturn::ArgIsNull)?
+            .iter()
+            .all(|item| FfiReturn::Ok == T::is_valid(*item))
+        {
+            return Err(FfiReturn::TrapRepresentation);
+        }
+
+        Ok(core::slice::from_raw_parts_mut(ptr.cast(), len))
+    }
+}
+
+impl<T: Transmute> CType<'_> for Vec<Transmutable<T, T::ReprC>> {
+    type ReprC = SliceRef<T::ReprC>;
+    type RustStore = Vec<Transmutable<T, T::ReprC>>;
+    type FfiStore = ();
+
+    fn into_repr_c(self, store: &mut Self::RustStore) -> Self::ReprC {
+        *store = self;
+
+        let (ptr, len) = (store.as_ptr(), store.len());
+        SliceRef::from_raw_parts(ptr.cast(), len)
+    }
+    unsafe fn try_from_repr_c(source: Self::ReprC, _: &mut ()) -> Result<Self> {
```

Comment:
that's true, but I'd rather not reallocate

Reply:
Uh, in this case i think calling `shrink_to_fit` is not suitable, because it call [`Allocator::shrink`](https://doc.rust-lang.org/src/core/alloc/mod.rs.html#319) inside which reallocates memory.

Reply:
Maybe I could go with `into_boxed_slice`

Reply:
The same problem actually ([source](https://doc.rust-lang.org/src/alloc/vec/mod.rs.html#1032))